src/nvim/ops.c

```diff
@@ -2524,6 +2528,61 @@ static void yank_copy_line(yankreg_T *reg, struct block_def *bd, long y_idx)
   *pnew = NUL;
 }
 
+static void yank_do_autocmd(oparg_T *oap, yankreg_T *reg)
+{
+  char buf[NUMBUFLEN+2];
+
+  if (!has_event(EVENT_TEXTYANKPOST)) {
+    // No autocommand was defined
+    return;
+  }
+
+  // set v:event to a dictionary with information about the yank
+  dict_T *dict = get_vim_var_dict(VV_EVENT);
+
+  // the yanked text
+  list_T *list = list_alloc();
+  for (linenr_T i = 0; i < reg->y_size; i++) {
+    list_append_string(list, reg->y_array[i], -1);
+  }
+  dict_add_list(dict, "regcontents", list);
+
+  // the register type
+  buf[0] = NUL;
+  buf[1] = NUL;
+  switch (reg->y_type) {
+    case MLINE:
+      buf[0] = 'V';
+      break;
+    case MCHAR:
+      buf[0] = 'v';
+      break;
+    case MBLOCK:
+      buf[0] = Ctrl_V;
+      snprintf(buf + 1, ARRAY_SIZE(buf) - 1, "%" PRId64,
+               (int64_t)(reg->y_width + 1));
+      break;
+    case MAUTO:
+      assert(false);
+  }
```

Comment:
It would be logical to move this code out of here to a function used both for `getregtype` and `yank_do_autocmd`:

```Patch
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index a0a686f..3693fba 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -10052,16 +10052,8 @@ static void f_getregtype(typval_T *argvars, typval_T *rettv)
   if (regname == 0)
     regname = '"';
 
-  buf[0] = NUL;
-  buf[1] = NUL;
-  switch (get_reg_type(regname, &reglen)) {
-  case MLINE: buf[0] = 'V'; break;
-  case MCHAR: buf[0] = 'v'; break;
-  case MBLOCK:
-    buf[0] = Ctrl_V;
-    sprintf((char *)buf + 1, "%" PRId64, (int64_t)(reglen + 1));
-    break;
-  }
+  op_get_register_type(NULL, regname, buf, sizeof(buf));
+
   rettv->v_type = VAR_STRING;
   rettv->vval.v_string = vim_strsave(buf);
 }
diff --git a/src/nvim/ops.c b/src/nvim/ops.c
index b1adc85..b3c8b1a 100644
--- a/src/nvim/ops.c
+++ b/src/nvim/ops.c
@@ -4626,39 +4626,57 @@ theend:
   return did_change;
 }
 
-/*
- * Return the type of a register.
- * Used for getregtype()
- * Returns MAUTO for error.
- */
-char_u get_reg_type(int regname, long *reglen)
+/// Write register type to a buffer
+///
+/// @param[in]  reg  Register to use.
+/// @param[in]  regname  Register name. Used in case reg is NULL.
+/// @param[out]  ret_buf  Buffer where results will be saved. If requested
+///                       register is not valid, ret_buf[0] is set to NUL, in
+///                       any other case it will be one of 'v', 'V' or Ctrl_V.
+/// @param[in]  ret_buf_len  Buffer length. Must be at least two characters.
+void op_get_register_type(const yankreg_T *reg, const int regname,
+                          char *const ret_buf, const size_t ret_buf_len)
+  FUNC_ATTR_NONNULL_ARG(3)
 {
-  switch (regname) {
-  case '%':                     /* file name */
-  case '#':                     /* alternate file name */
-  case '=':                     /* expression */
-  case ':':                     /* last command line */
-  case '/':                     /* last search-pattern */
-  case '.':                     /* last inserted text */
-  case Ctrl_F:                  /* Filename under cursor */
-  case Ctrl_P:                  /* Path under cursor, expand via "path" */
-  case Ctrl_W:                  /* word under cursor */
-  case Ctrl_A:                  /* WORD (mnemonic All) under cursor */
-  case '_':                     /* black hole: always empty */
-    return MCHAR;
+  assert(buf_len > 1);
+  if (reg == NULL) {
+    switch (regname) {
+      case '%':     // file name
+      case '#':     // alternate file name
+      case '=':     // expression
+      case ':':     // last command line
+      case '/':     // last search-pattern
+      case '.':     // last inserted text
+      case Ctrl_F:  // filename under cursor
+      case Ctrl_P:  // path under cursor, expand via "path"
+      case Ctrl_W:  // word under cursor
+      case Ctrl_A:  // WORD (mnemonic All) under cursor
+      case '_': {   // black hole: always empty
+        ret_buf[0] = 'v';
+        ret_buf[1] = NUL;
+        return;
+      }
+      default: {
+        if (regname != NUL && !valid_yank_reg(regname, false)) {
+          reg = get_yank_register(regname, YREG_PASTE);
+        }
+        break;
+      }
+    }
   }
-
-  if (regname != NUL && !valid_yank_reg(regname, false))
-    return MAUTO;
-
-  yankreg_T *reg = get_yank_register(regname, YREG_PASTE);
-
   if (reg->y_array != NULL) {
-    if (reglen != NULL && reg->y_type == MBLOCK)
-      *reglen = reg->y_width;
-    return reg->y_type;
+    switch (reg->y_type) {
+      case MLINE: ret_buf[0] = 'V'; ret_buf[1] = NUL; break;
+      case MCHAR: ret_buf[0] = 'v'; ret_buf[1] = NUL; break;
+      case MBLOCK: {
+        snprintf(ret_buf, ret_buf_len, CTRL_V_STR "%" PRIdLINENR,
+                 reg->y_width + 1);
+        break;
+      }
+    }
+    return;
   }
-  return MAUTO;
+  ret_buf[0] = NUL;
 }
 
 /// When `flags` has `kGRegList` return a list with text `s`.
diff --git a/src/nvim/pos.h b/src/nvim/pos.h
index 7071df5..f517337 100644
--- a/src/nvim/pos.h
+++ b/src/nvim/pos.h
@@ -1,7 +1,11 @@
 #ifndef NVIM_POS_H
 #define NVIM_POS_H
 
-typedef long linenr_T;         // line number type
+/// Line number type
+typedef long linenr_T;
+/// Format used to print values which have linenr_T type
+#define PRIdLINENR "ld"
+
 typedef int colnr_T;           // column number type
 
 #define MAXLNUM 0x7fffffff     // maximum (invalid) line number
```